test/get-basic-auth-header.js

```diff
@@ -0,0 +1,67 @@
+const test = require('ava');
+const getBasicAuthHeader = require('../source/proxies/get-basic-auth-header')
+
+test.serial('Encodes username and password as they appear in the URL', t => {
+    const cases = [
+        {
+            url: new URL('https://user;name:p+ssword@example.com'),
+            username: 'user;name',
+            password: 'p+ssword'
+        },
+
+        {
+            url: new URL('https://user:@example.com'),
+            username: 'user',
+            password: ''
+        },
+
+        {
+            url: new URL('https://:password@example.com'),
+            username: '',
+            password: 'password'
+        }
+    ];
+
+    for (const { url, username, password } of cases) {
+        const expectedHeader = getExpectedHeader(username, password);
+        const actualHeader = getBasicAuthHeader(url);
+
+        t.is(actualHeader, expectedHeader, url.toString());
+    }
+});
+
+function getExpectedHeader(username, password) {
+    const base64Part = Buffer
+        .from(`${username}:${password}`)
+        .toString('base64');
+
+    return `Basic ${base64Part}`;
+}
+
+test.serial('Throws when username/password contains lone %', t => {
+    // According to RFC 1738, username and password
+    // cannot contain % that is not followed by 2 hexadecimal
+    // digits
+    //
+    // See user, password, and uchar here:
+    // https://datatracker.ietf.org/doc/html/rfc1738#section-5
```

Comment:
>    The user name (and password), if present, are followed by a
>    commercial at-sign "@". Within the user and password field, any ":",
>    "@", or "/" must be encoded.

Therefore the decoding has to be done on the server, because according to the Basic auth RFC the username can't contain a colon.

Reply:
Could you clarify what does this comment address? In the case of my proxy, username does not contain ':' but a ';'

Reply:
Interesting point. I've quoted the RFC 1738 for URL, but I've realized that it's not necessary for the username and password in the Basic auth scheme to follow the same format as in the URL

[RFC for basic auth](https://datatracker.ietf.org/doc/html/rfc7617) does not reference RFC 1738. It also has this paragraph:

> Furthermore, a user-id containing a colon character is invalid, as
> the first colon in a user-pass string separates user-id and password
> from one another; text after the first colon is part of the password.
> **User-ids containing colons cannot be encoded in user-pass strings**.

`user-id` and `user-pass` are define here:

``` 
To receive authorization, the client

1. obtains the user-id and password from the user,
2. constructs the user-pass by concatenating the user-id, a single colon (":") character, and the password,
```

Do you think this suggests that `user-pass` is not supposed to have any kind of encoding?

Reply:
My guess is that the two RFCs aren't related at all. The URL spec says that the user field can contain a colon. The Basic spec says it can't. The URL does not enforce Basic usage nor the Basic spec enforces URL.

I'd prefer to leave things as-is.

Reply:
As someone who spent time tracing a cryptic error related to proxy, I would like such URLs to be handled :)

I think there's a relatively simple way to do that:

We can use `percentDecodeString()` from library [jsdom/whatwg-url](https://github.com/jsdom/whatwg-url) to percent-decode `URL.username` and `URL.password`, and use the plain values in Basic auth scheme

The code that computes the value of `Proxy-Authorization` would look like this:

```ts
const url = new URL(options.proxyUrl)
const username = percentDecodeString(url.username)
const password = percentDecodeString(url.password)

const credentials = `${username}:${password}`
const credentialsBase64 = Buffer.from(credentials).toString('base64')

const proxyAuthorization = `Basic ${credentialsBase64}`
```

If you think this is good idea, I can submit PR with tests

### Details

#### URL format

First, a bit of nitpicking, but we need to define what format of URL we expect in `proxyUrl`. There are multiple, slightly different URL formats. The code already uses `new URL()` for parsing, which treats URL as being in [WHATWG URL format](https://url.spec.whatwg.org). This is commonly used format, it tries to reflect actual browsers' behavior, and is widely used in Node.js

So we say that `proxyUrl` must be a string of URL in WHATWG URL format

#### Getting plain username/password

Next, we need to obtain plain username/password from the parsed URL

In WHATWG URL, `url.username` and `url.password` return username/password encoded using exotic percent-encoding. [Here is the place in spec](https://url.spec.whatwg.org/#authority-state) where username/password are parsed. You can see in step 1.4.2 that username/password are [UTF-8 percent-encoded](https://url.spec.whatwg.org/#utf-8-percent-encode) using the [userinfo percent-encode set](https://url.spec.whatwg.org/#userinfo-percent-encode-set)

That set defines what characters to percent-encode. `%` is not among them, so `new URL()` leaves percents in username/password as they are:

```
> const url = new URL('https://us%r:p%ssword@example.com')
> url.username
'us%r'
> url.password
'p%ssword'
```

That's why we cannot use `decodeURIComponent()`: it breaks with lone percents

```
> decodeURIComponent('us%r')
Uncaught URIError: URI malformed
    at decodeURIComponent (<anonymous>)
```

Instead, we need to use the same percent-enconding/decoding algorithm as WHATWG URL. That's exactly what `percentDecodeString()` does. It is implemented according to [percent-decode in the spec](https://url.spec.whatwg.org/#string-percent-decode)